Remove WheelBuilder's dependence on PipSession

cjerdonek opened the pull request:
This removes the `WheelBuilder` class's dependence on `PipSession`. It turns out `WheelBuilder` doesn't actually need to call the full `unpack_url()` -- only `unpack_file_url()`, which doesn't require network access. This simplifies some things.

I also added some minimal type-checking to the `install` and `wheel` command modules so I could have the benefit of mypy type-checking the calls to `WheelBuilder.build()`, which is what this PR changes.

cjerdonek commented:
As a follow-on, I think `unpack_file_url()` should be moved out of `download.py` (since it doesn't do any downloading or network access -- just unpacking on the file system). That would let us remove `wheel.py`'s import dependency on `download.py`.

cjerdonek commented:
Also, to make reviewing easier, below is the source for `unpack_url()`. When `WheelBuilder` calls `unpack_url()`, the link is a file URL (to a local wheel) and not a VCS URL: https://github.com/pypa/pip/blob/2bfafc9b70f76ec0b019c96311cb437e739b21e2/src/pip/_internal/download.py#L1057-L1079

chrahunt commented on src/pip/_internal/commands/install.py:
```diff
@@ -30,9 +34,16 @@
+
+    from pip._internal.req.req_install import InstallRequirement
+
 logger = logging.getLogger(__name__)
```
I would put two empty lines after imports and before body contents.

cjerdonek replied:
Done in the other file as well.

pradyunsg commented on src/pip/_internal/commands/install.py:
```diff
@@ -1,3 +1,7 @@
+
+# The following comment should be removed at some point in the future.
+# mypy: strict-optional=False
```
nit: It'd probably help to flag/add comments to lines that require this, especially if you looked into this and have some insight into what could be done.

pradyunsg commented on src/pip/_internal/commands/install.py:
```diff
@@ -58,7 +74,7 @@ def build_wheels(builder, pep517_requirements, legacy_requirements, session):
     build_failures = builder.build(
         pep517_requirements,
-        session=session, autobuilding=True
+        autobuilding=True
```
nit: Add trailing commas while we're changing this?

pradyunsg commented on src/pip/_internal/commands/install.py:
```diff
@@ -67,7 +83,7 @@ def build_wheels(builder, pep517_requirements, legacy_requirements, session):
         builder.build(
             legacy_requirements,
-            session=session, autobuilding=True
+            autobuilding=True
```
nit: Add trailing commas while we're changing this?

cjerdonek commented:
Thanks for the quick reviews, @pradyunsg and @chrahunt! I addressed all your comments, which were all great and I agreed with.

> nit: Add trailing commas while we're changing this?

Is it possible to have a linter check this when it overlaps with the change?

pradyunsg commented:
> Is it possible to have a linter check this when it overlaps with the change?

I investigated for a bit -- there's [flake8-commas](https://github.com/PyCQA/flake8-commas) and [add-trailing-comma](https://github.com/asottile/add-trailing-comma).

It should be do-able. FWIW, it also makes sense to adopt `pre-commit` at some point for the linting workload.

cjerdonek commented:
> As a follow-on, I think `unpack_file_url()` should be moved out of `download.py`

I created issue #6861 for this.
